{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "this notebook best viewed here:  https://nbviewer.jupyter.org/github/brandonlind/BURT/blob/master/001_assemblies.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pythonimports import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# denovo_all_samps\n",
    "create an assembly using all samples across all species\n",
    "\n",
    "."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### make some dirs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIR = makedir('/gpfs_fs/home/eckertlab/BURT/denovo_all_samps')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "fqdir = makedir('/gpfs_fs/home/eckertlab/BURT/fastqs_for_dDocent')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### read in datatable, creats symlinks to fqdir "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fastq</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>species</th>\n",
       "      <th>state</th>\n",
       "      <th>ind</th>\n",
       "      <th>popn</th>\n",
       "      <th>spp_popn</th>\n",
       "      <th>run</th>\n",
       "      <th>lib</th>\n",
       "      <th>seq_center</th>\n",
       "      <th>md5</th>\n",
       "      <th>RGPL</th>\n",
       "      <th>RGSM</th>\n",
       "      <th>RGLB</th>\n",
       "      <th>RGID</th>\n",
       "      <th>flowcell_lane</th>\n",
       "      <th>RGPU</th>\n",
       "      <th>num_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>T-AR-2-7</td>\n",
       "      <td>T</td>\n",
       "      <td>AR</td>\n",
       "      <td>7</td>\n",
       "      <td>T-AR-2</td>\n",
       "      <td>AR-2</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>1ae81d40ba5b1497e7ba9f5733c494b4</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>T-AR-2-7</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>T-AR-2-7.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.T-AR-2-7</td>\n",
       "      <td>1277586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>T-TX-1-19</td>\n",
       "      <td>T</td>\n",
       "      <td>TX</td>\n",
       "      <td>19</td>\n",
       "      <td>T-TX-1</td>\n",
       "      <td>TX-1</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>eb991e8800c31b632dfc0f4279e9e44a</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>T-TX-1-19</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>T-TX-1-19.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.T-TX-1-19</td>\n",
       "      <td>2798860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>G-SC-1-3</td>\n",
       "      <td>G</td>\n",
       "      <td>SC</td>\n",
       "      <td>3</td>\n",
       "      <td>G-SC-1</td>\n",
       "      <td>SC-1</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>e5cf36460b50c6467a49f4b157bb16c1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>G-SC-1-3</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>G-SC-1-3.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.G-SC-1-3</td>\n",
       "      <td>2312577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>P-LA-2-10</td>\n",
       "      <td>P</td>\n",
       "      <td>LA</td>\n",
       "      <td>10</td>\n",
       "      <td>P-LA-2</td>\n",
       "      <td>LA-2</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>a1fab3ce615d708f53eb836f13fbba7c</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>P-LA-2-10</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>P-LA-2-10.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.P-LA-2-10</td>\n",
       "      <td>1583884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>T-AL-1-2</td>\n",
       "      <td>T</td>\n",
       "      <td>AL</td>\n",
       "      <td>2</td>\n",
       "      <td>T-AL-1</td>\n",
       "      <td>AL-1</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>16efa4d02874e6f5cf747e40bdbfb0dd</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>T-AL-1-2</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>T-AL-1-2.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.T-AL-1-2</td>\n",
       "      <td>1321261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1704</th>\n",
       "      <td>/home/cfriedline/eckertlab/Novogene/burt/demul...</td>\n",
       "      <td>T-MS-4-15</td>\n",
       "      <td>T</td>\n",
       "      <td>MS</td>\n",
       "      <td>15</td>\n",
       "      <td>T-MS-4</td>\n",
       "      <td>MS-4</td>\n",
       "      <td>Novogene1</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>Novogene</td>\n",
       "      <td>2340098d47990e9fb05d65282c30868c</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>T-MS-4-15</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>T-MS-4-15.Novogene1.BURT_13</td>\n",
       "      <td>HKFKHBBXX.5</td>\n",
       "      <td>HKFKHBBXX.5.T-MS-4-15</td>\n",
       "      <td>848933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1705</th>\n",
       "      <td>/home/cfriedline/eckertlab/Novogene/burt/demul...</td>\n",
       "      <td>G-GA-2-12</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>12</td>\n",
       "      <td>G-GA-2</td>\n",
       "      <td>GA-2</td>\n",
       "      <td>Novogene1</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>Novogene</td>\n",
       "      <td>31fe9cac2482474e729677bbe38c871e</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>G-GA-2-12</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>G-GA-2-12.Novogene1.BURT_13</td>\n",
       "      <td>HKFKHBBXX.5</td>\n",
       "      <td>HKFKHBBXX.5.G-GA-2-12</td>\n",
       "      <td>4920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1706</th>\n",
       "      <td>/home/cfriedline/eckertlab/Novogene/burt/demul...</td>\n",
       "      <td>P-GA-5-10</td>\n",
       "      <td>P</td>\n",
       "      <td>GA</td>\n",
       "      <td>10</td>\n",
       "      <td>P-GA-5</td>\n",
       "      <td>GA-5</td>\n",
       "      <td>Novogene1</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>Novogene</td>\n",
       "      <td>c13f6689db7b0e5ed01f9906c681c4c9</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>P-GA-5-10</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>P-GA-5-10.Novogene1.BURT_13</td>\n",
       "      <td>HKFKHBBXX.5</td>\n",
       "      <td>HKFKHBBXX.5.P-GA-5-10</td>\n",
       "      <td>35709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1707</th>\n",
       "      <td>/home/cfriedline/eckertlab/Novogene/burt/demul...</td>\n",
       "      <td>E-MS-1-5</td>\n",
       "      <td>E</td>\n",
       "      <td>MS</td>\n",
       "      <td>5</td>\n",
       "      <td>E-MS-1</td>\n",
       "      <td>MS-1</td>\n",
       "      <td>Novogene1</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>Novogene</td>\n",
       "      <td>247292352701138aa0b24a82a56e6382</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>E-MS-1-5</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>E-MS-1-5.Novogene1.BURT_13</td>\n",
       "      <td>HKFKHBBXX.5</td>\n",
       "      <td>HKFKHBBXX.5.E-MS-1-5</td>\n",
       "      <td>1488612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1708</th>\n",
       "      <td>/home/cfriedline/eckertlab/Novogene/burt/demul...</td>\n",
       "      <td>T-LA-3-12</td>\n",
       "      <td>T</td>\n",
       "      <td>LA</td>\n",
       "      <td>12</td>\n",
       "      <td>T-LA-3</td>\n",
       "      <td>LA-3</td>\n",
       "      <td>Novogene1</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>Novogene</td>\n",
       "      <td>c3a71fccbf18ece237525aa962b7f775</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>T-LA-3-12</td>\n",
       "      <td>BURT_13</td>\n",
       "      <td>T-LA-3-12.Novogene1.BURT_13</td>\n",
       "      <td>HKFKHBBXX.5</td>\n",
       "      <td>HKFKHBBXX.5.T-LA-3-12</td>\n",
       "      <td>928345</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1709 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  fastq sample_name species  \\\n",
       "0     /home/cfriedline/eckertlab/projects/burt/seq/1...    T-AR-2-7       T   \n",
       "1     /home/cfriedline/eckertlab/projects/burt/seq/1...   T-TX-1-19       T   \n",
       "2     /home/cfriedline/eckertlab/projects/burt/seq/1...    G-SC-1-3       G   \n",
       "3     /home/cfriedline/eckertlab/projects/burt/seq/1...   P-LA-2-10       P   \n",
       "4     /home/cfriedline/eckertlab/projects/burt/seq/1...    T-AL-1-2       T   \n",
       "...                                                 ...         ...     ...   \n",
       "1704  /home/cfriedline/eckertlab/Novogene/burt/demul...   T-MS-4-15       T   \n",
       "1705  /home/cfriedline/eckertlab/Novogene/burt/demul...   G-GA-2-12       G   \n",
       "1706  /home/cfriedline/eckertlab/Novogene/burt/demul...   P-GA-5-10       P   \n",
       "1707  /home/cfriedline/eckertlab/Novogene/burt/demul...    E-MS-1-5       E   \n",
       "1708  /home/cfriedline/eckertlab/Novogene/burt/demul...   T-LA-3-12       T   \n",
       "\n",
       "     state  ind    popn spp_popn        run      lib seq_center  \\\n",
       "0       AR    7  T-AR-2     AR-2     160520    Burt0       NARF   \n",
       "1       TX   19  T-TX-1     TX-1     160520    Burt0       NARF   \n",
       "2       SC    3  G-SC-1     SC-1     160520    Burt0       NARF   \n",
       "3       LA   10  P-LA-2     LA-2     160520    Burt0       NARF   \n",
       "4       AL    2  T-AL-1     AL-1     160520    Burt0       NARF   \n",
       "...    ...  ...     ...      ...        ...      ...        ...   \n",
       "1704    MS   15  T-MS-4     MS-4  Novogene1  BURT_13   Novogene   \n",
       "1705    GA   12  G-GA-2     GA-2  Novogene1  BURT_13   Novogene   \n",
       "1706    GA   10  P-GA-5     GA-5  Novogene1  BURT_13   Novogene   \n",
       "1707    MS    5  E-MS-1     MS-1  Novogene1  BURT_13   Novogene   \n",
       "1708    LA   12  T-LA-3     LA-3  Novogene1  BURT_13   Novogene   \n",
       "\n",
       "                                   md5      RGPL       RGSM     RGLB  \\\n",
       "0     1ae81d40ba5b1497e7ba9f5733c494b4  ILLUMINA   T-AR-2-7    Burt0   \n",
       "1     eb991e8800c31b632dfc0f4279e9e44a  ILLUMINA  T-TX-1-19    Burt0   \n",
       "2     e5cf36460b50c6467a49f4b157bb16c1  ILLUMINA   G-SC-1-3    Burt0   \n",
       "3     a1fab3ce615d708f53eb836f13fbba7c  ILLUMINA  P-LA-2-10    Burt0   \n",
       "4     16efa4d02874e6f5cf747e40bdbfb0dd  ILLUMINA   T-AL-1-2    Burt0   \n",
       "...                                ...       ...        ...      ...   \n",
       "1704  2340098d47990e9fb05d65282c30868c  ILLUMINA  T-MS-4-15  BURT_13   \n",
       "1705  31fe9cac2482474e729677bbe38c871e  ILLUMINA  G-GA-2-12  BURT_13   \n",
       "1706  c13f6689db7b0e5ed01f9906c681c4c9  ILLUMINA  P-GA-5-10  BURT_13   \n",
       "1707  247292352701138aa0b24a82a56e6382  ILLUMINA   E-MS-1-5  BURT_13   \n",
       "1708  c3a71fccbf18ece237525aa962b7f775  ILLUMINA  T-LA-3-12  BURT_13   \n",
       "\n",
       "                             RGID flowcell_lane                   RGPU  \\\n",
       "0           T-AR-2-7.160520.Burt0   C7WRGACXX.1   C7WRGACXX.1.T-AR-2-7   \n",
       "1          T-TX-1-19.160520.Burt0   C7WRGACXX.1  C7WRGACXX.1.T-TX-1-19   \n",
       "2           G-SC-1-3.160520.Burt0   C7WRGACXX.1   C7WRGACXX.1.G-SC-1-3   \n",
       "3          P-LA-2-10.160520.Burt0   C7WRGACXX.1  C7WRGACXX.1.P-LA-2-10   \n",
       "4           T-AL-1-2.160520.Burt0   C7WRGACXX.1   C7WRGACXX.1.T-AL-1-2   \n",
       "...                           ...           ...                    ...   \n",
       "1704  T-MS-4-15.Novogene1.BURT_13   HKFKHBBXX.5  HKFKHBBXX.5.T-MS-4-15   \n",
       "1705  G-GA-2-12.Novogene1.BURT_13   HKFKHBBXX.5  HKFKHBBXX.5.G-GA-2-12   \n",
       "1706  P-GA-5-10.Novogene1.BURT_13   HKFKHBBXX.5  HKFKHBBXX.5.P-GA-5-10   \n",
       "1707   E-MS-1-5.Novogene1.BURT_13   HKFKHBBXX.5   HKFKHBBXX.5.E-MS-1-5   \n",
       "1708  T-LA-3-12.Novogene1.BURT_13   HKFKHBBXX.5  HKFKHBBXX.5.T-LA-3-12   \n",
       "\n",
       "      num_reads  \n",
       "0       1277586  \n",
       "1       2798860  \n",
       "2       2312577  \n",
       "3       1583884  \n",
       "4       1321261  \n",
       "...         ...  \n",
       "1704     848933  \n",
       "1705       4920  \n",
       "1706      35709  \n",
       "1707    1488612  \n",
       "1708     928345  \n",
       "\n",
       "[1709 rows x 18 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_table(op.join(op.dirname(DIR), 'seq/burt_fastq_data.txt'))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1135"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "luni(df['sample_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1709"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nrow(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### concatenate samps sequenced > 1 (or symlink) into fastqdir for dDocent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({True: 1709})"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make sure all of the fastq data exists\n",
    "counts = Counter()\n",
    "for fq in df['fastq']:\n",
    "    counts[op.exists(fq)] += 1\n",
    "counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1135, 1709)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# are samples sequenced > 1. A: YES, therefore I need to concat files\n",
    "luni(df['sample_name']), len(df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fastq</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>species</th>\n",
       "      <th>state</th>\n",
       "      <th>ind</th>\n",
       "      <th>popn</th>\n",
       "      <th>spp_popn</th>\n",
       "      <th>run</th>\n",
       "      <th>lib</th>\n",
       "      <th>seq_center</th>\n",
       "      <th>md5</th>\n",
       "      <th>RGPL</th>\n",
       "      <th>RGSM</th>\n",
       "      <th>RGLB</th>\n",
       "      <th>RGID</th>\n",
       "      <th>flowcell_lane</th>\n",
       "      <th>RGPU</th>\n",
       "      <th>num_reads</th>\n",
       "      <th>dDocent_sampname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>T-AR-2-7</td>\n",
       "      <td>T</td>\n",
       "      <td>AR</td>\n",
       "      <td>7</td>\n",
       "      <td>T-AR-2</td>\n",
       "      <td>AR-2</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>1ae81d40ba5b1497e7ba9f5733c494b4</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>T-AR-2-7</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>T-AR-2-7.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.T-AR-2-7</td>\n",
       "      <td>1277586</td>\n",
       "      <td>T-AR-2_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>T-TX-1-19</td>\n",
       "      <td>T</td>\n",
       "      <td>TX</td>\n",
       "      <td>19</td>\n",
       "      <td>T-TX-1</td>\n",
       "      <td>TX-1</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>eb991e8800c31b632dfc0f4279e9e44a</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>T-TX-1-19</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>T-TX-1-19.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.T-TX-1-19</td>\n",
       "      <td>2798860</td>\n",
       "      <td>T-TX-1_19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>G-SC-1-3</td>\n",
       "      <td>G</td>\n",
       "      <td>SC</td>\n",
       "      <td>3</td>\n",
       "      <td>G-SC-1</td>\n",
       "      <td>SC-1</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>e5cf36460b50c6467a49f4b157bb16c1</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>G-SC-1-3</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>G-SC-1-3.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.G-SC-1-3</td>\n",
       "      <td>2312577</td>\n",
       "      <td>G-SC-1_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>P-LA-2-10</td>\n",
       "      <td>P</td>\n",
       "      <td>LA</td>\n",
       "      <td>10</td>\n",
       "      <td>P-LA-2</td>\n",
       "      <td>LA-2</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>a1fab3ce615d708f53eb836f13fbba7c</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>P-LA-2-10</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>P-LA-2-10.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.P-LA-2-10</td>\n",
       "      <td>1583884</td>\n",
       "      <td>P-LA-2_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>/home/cfriedline/eckertlab/projects/burt/seq/1...</td>\n",
       "      <td>T-AL-1-2</td>\n",
       "      <td>T</td>\n",
       "      <td>AL</td>\n",
       "      <td>2</td>\n",
       "      <td>T-AL-1</td>\n",
       "      <td>AL-1</td>\n",
       "      <td>160520</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>NARF</td>\n",
       "      <td>16efa4d02874e6f5cf747e40bdbfb0dd</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>T-AL-1-2</td>\n",
       "      <td>Burt0</td>\n",
       "      <td>T-AL-1-2.160520.Burt0</td>\n",
       "      <td>C7WRGACXX.1</td>\n",
       "      <td>C7WRGACXX.1.T-AL-1-2</td>\n",
       "      <td>1321261</td>\n",
       "      <td>T-AL-1_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               fastq sample_name species  \\\n",
       "0  /home/cfriedline/eckertlab/projects/burt/seq/1...    T-AR-2-7       T   \n",
       "1  /home/cfriedline/eckertlab/projects/burt/seq/1...   T-TX-1-19       T   \n",
       "2  /home/cfriedline/eckertlab/projects/burt/seq/1...    G-SC-1-3       G   \n",
       "3  /home/cfriedline/eckertlab/projects/burt/seq/1...   P-LA-2-10       P   \n",
       "4  /home/cfriedline/eckertlab/projects/burt/seq/1...    T-AL-1-2       T   \n",
       "\n",
       "  state  ind    popn spp_popn     run    lib seq_center  \\\n",
       "0    AR    7  T-AR-2     AR-2  160520  Burt0       NARF   \n",
       "1    TX   19  T-TX-1     TX-1  160520  Burt0       NARF   \n",
       "2    SC    3  G-SC-1     SC-1  160520  Burt0       NARF   \n",
       "3    LA   10  P-LA-2     LA-2  160520  Burt0       NARF   \n",
       "4    AL    2  T-AL-1     AL-1  160520  Burt0       NARF   \n",
       "\n",
       "                                md5      RGPL       RGSM   RGLB  \\\n",
       "0  1ae81d40ba5b1497e7ba9f5733c494b4  ILLUMINA   T-AR-2-7  Burt0   \n",
       "1  eb991e8800c31b632dfc0f4279e9e44a  ILLUMINA  T-TX-1-19  Burt0   \n",
       "2  e5cf36460b50c6467a49f4b157bb16c1  ILLUMINA   G-SC-1-3  Burt0   \n",
       "3  a1fab3ce615d708f53eb836f13fbba7c  ILLUMINA  P-LA-2-10  Burt0   \n",
       "4  16efa4d02874e6f5cf747e40bdbfb0dd  ILLUMINA   T-AL-1-2  Burt0   \n",
       "\n",
       "                     RGID flowcell_lane                   RGPU  num_reads  \\\n",
       "0   T-AR-2-7.160520.Burt0   C7WRGACXX.1   C7WRGACXX.1.T-AR-2-7    1277586   \n",
       "1  T-TX-1-19.160520.Burt0   C7WRGACXX.1  C7WRGACXX.1.T-TX-1-19    2798860   \n",
       "2   G-SC-1-3.160520.Burt0   C7WRGACXX.1   C7WRGACXX.1.G-SC-1-3    2312577   \n",
       "3  P-LA-2-10.160520.Burt0   C7WRGACXX.1  C7WRGACXX.1.P-LA-2-10    1583884   \n",
       "4   T-AL-1-2.160520.Burt0   C7WRGACXX.1   C7WRGACXX.1.T-AL-1-2    1321261   \n",
       "\n",
       "  dDocent_sampname  \n",
       "0         T-AR-2_7  \n",
       "1        T-TX-1_19  \n",
       "2         G-SC-1_3  \n",
       "3        P-LA-2_10  \n",
       "4         T-AL-1_2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert samp names into something dDocent expects, will make it easier later\n",
    "df['dDocent_sampname'] = [f'{pop}_{samp}' for (pop,samp) in zip(df['popn'], df['ind'])]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure each sampname only has one \"_\"\n",
    "for samp in df['dDocent_sampname']:\n",
    "    assert len(samp.split(\"_\")) == 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get files for each sample (len(list) = count)\n",
    "sampfiles = {}\n",
    "for row in df.index:\n",
    "    samp = df.loc[row, 'dDocent_sampname']\n",
    "    if samp not in sampfiles:\n",
    "        sampfiles[samp] = []\n",
    "    sampfiles[samp].append(df.loc[row, 'fastq'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create commands to concatenate samps sequenced > 1, otherwise just use a symlink to save space\n",
    "def getdst(samp, num):\n",
    "    # if symlink, it says unzipped (.fq suffix), but the src file is gzipped\n",
    "    if num == 1:\n",
    "        return op.join(fqdir, f'{samp}.F.fq.gz')\n",
    "    else:\n",
    "        return op.join(fqdir, f'{samp}.F.fq')\n",
    "\n",
    "catdir = makedir(op.join(DIR, 'catfiles'))\n",
    "commandfile = op.join(catdir, 'concatenate_fastqs.txt')\n",
    "cmds = []\n",
    "dsts = []\n",
    "for samp,lst in sampfiles.items():\n",
    "    if len(lst) == 1:\n",
    "        dst = getdst(samp, 1)\n",
    "        try:\n",
    "            os.symlink(lst[0], dst)\n",
    "        except:\n",
    "            print(\"couldn't symlink %s\" % src)\n",
    "    else:\n",
    "        dst = getdst(samp, 0)\n",
    "        dsts.append(dst)\n",
    "        cmds.append('zcat %s > %s' % (' '.join(lst), dst))\n",
    "with open(commandfile, 'w') as o:\n",
    "    o.write(\"%s\" % '\\n'.join(cmds))\n",
    "\n",
    "# go gzip the concatenated files\n",
    "gzipfile = op.join(catdir, 'gzip_concatenate_fastqs.txt')\n",
    "cmds = []\n",
    "for dst in dsts:\n",
    "    cmds.append(f'gzip -f {dst}')\n",
    "with open(gzipfile, 'w') as o:\n",
    "    o.write(\"%s\" % '\\n'.join(cmds))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "# I did this in terminal\n",
    "ssh godel97\n",
    "cd ~/eckertlab/BURT/denovo_all_samps/catfiles\n",
    "cat concatenate_fastqs.txt | parallel -j 64 --progress\n",
    "\n",
    "cat gzip_concatenate_fastqs.txt | parallel -j 64 --progress\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# do the number of files match the number of samples? A: YES\n",
    "len(fs(fqdir)) == luni(df['dDocent_sampname'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gpfs_fs/home/eckertlab/BURT/denovo_all_samps/catfiles'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "catdir = makedir(op.join(DIR, 'catfiles'))\n",
    "catdir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gpfs_fs/home/eckertlab/BURT/denovo_all_samps'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DIR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Do you want to perform an assembly?\n",
    "Type yes or no and press [ENTER].\n",
    "yes\n",
    "What type of assembly would you like to perform?  Enter SE for single end, PE for paired-end, RPE for paired-end sequencing for RAD protocols with random shearing, or OL for paired-end sequencing that has substantial overlap.\n",
    "Then press [ENTER]\n",
    "SE\n",
    "Reads will be assembled with Rainbow\n",
    "CD-HIT will cluster reference sequences by similarity. The -c parameter (% similarity to cluster) may need to be changed for your taxa.\n",
    "Would you like to enter a new c parameter now? Type yes or no and press [ENTER]\n",
    "yes\n",
    "Please enter new value for c. Enter in decimal form (For 90%, enter 0.9)\n",
    "0.85\n",
    "Do you want to map reads?  Type yes or no and press [ENTER]\n",
    "no\n",
    "Mapping will not be performed\n",
    "Do you want to use FreeBayes to call SNPs?  Please type yes or no and press [ENTER]\n",
    "no\n",
    "\n",
    "Please enter your email address.  dDocent will email you when it is finished running.\n",
    "Don't worry; dDocent has no financial need to sell your email address to spammers.\n",
    "lindb@vcu.edu\n",
    "\n",
    "\n",
    "dDocent will require input during the assembly stage.  Please wait until prompt says it is safe to move program to the background.\n",
    "\n",
    "Trimming reads and simultaneously assembling reference sequences\n",
    "\n",
    "[1] 1:bash*                                                                                                                                                                                                                                                              godel97"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### start dDocent in a tmux window on godel97\n",
    "configtext = '''Number of Processors\n",
    "64\n",
    "Maximum Memory\n",
    "600\n",
    "Trimming\n",
    "yes\n",
    "Assembly?\n",
    "yes\n",
    "Type_of_Assembly\n",
    "SE\n",
    "Clustering_Similarity%\n",
    "0.85\n",
    "Minimum within individual coverage level to include a read for assembly (K1)\n",
    "4\n",
    "Minimum number of individuals a read must be present in to include for assembly (K2)\n",
    "113\n",
    "Mapping_Reads?\n",
    "yes\n",
    "Mapping_Match_Value\n",
    "1\n",
    "Mapping_MisMatch_Value\n",
    "4\n",
    "Mapping_GapOpen_Penalty\n",
    "6\n",
    "Calling_SNPs?\n",
    "yes\n",
    "Email\n",
    "lindb@vcu.edu'''\n",
    "configfile = op.join(DIR, 'all_samps_de_novo_assembly.config')\n",
    "with open(configfile, 'w') as o:\n",
    "    o.write(\"%s\" % configtext)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gpfs_fs/home/eckertlab/BURT/denovo_all_samps/all_samps_de_novo_assembly.config'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "configfile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### output from run with no config, I said don't map or call SNPs, so using config above to run on this output\n",
    "```bash\n",
    "\n",
    "                       Number of Unique Sequences with More than X Coverage (Counted within individuals)\n",
    "\n",
    "  1.2e+08 +---------------------------------------------------------------------------------------------------------+\n",
    "          |           +           +          +           +           +           +          +           +           |\n",
    "          |                                                                                                         |\n",
    "          |*                                                                                                        |\n",
    "    1e+08 |*+                                                                                                     +-|\n",
    "          | *                                                                                                       |\n",
    "          | *                                                                                                       |\n",
    "          | *                                                                                                       |\n",
    "    8e+07 |-+*                                                                                                    +-|\n",
    "          |  *                                                                                                      |\n",
    "          |  *                                                                                                      |\n",
    "    6e+07 |-+ *                                                                                                   +-|\n",
    "          |   *                                                                                                     |\n",
    "          |    *                                                                                                    |\n",
    "          |    *                                                                                                    |\n",
    "    4e+07 |-+  *                                                                                                  +-|\n",
    "          |     *                                                                                                   |\n",
    "          |     ***                                                                                                 |\n",
    "          |        *                                                                                                |\n",
    "    2e+07 |-+       **                                                                                            +-|\n",
    "          |                                                                                                         |\n",
    "          |           ******                                                                                        |\n",
    "          |           +     ***********************      +           +           +          +           +           |\n",
    "        0 +---------------------------------------------------------------------------------------------------------+\n",
    "          2           4           6          8           10          12          14         16          18          20\n",
    "                                                           Coverage\n",
    "\n",
    "Please choose data cutoff.  In essence, you are picking a minimum (within individual) coverage level for a read (allele) to be used in the reference assembly\n",
    "4\n",
    "\n",
    "\n",
    "                                 Number of Unique Sequences present in more than X Individuals\n",
    "\n",
    "   900000 +---------------------------------------------------------------------------------------------------------+\n",
    "          |                 +                +                 +                 +                +                 |\n",
    "          |                                                                                                         |\n",
    "   800000 |-+                                                                                                     +-|\n",
    "          |                                                                                                         |\n",
    "   700000 |-+                                                                                                     +-|\n",
    "          |*                                                                                                        |\n",
    "          |*                                                                                                        |\n",
    "   600000 |*+                                                                                                     +-|\n",
    "          |*                                                                                                        |\n",
    "   500000 |*+                                                                                                     +-|\n",
    "          |*                                                                                                        |\n",
    "          |*                                                                                                        |\n",
    "   400000 |*+                                                                                                     +-|\n",
    "          |*                                                                                                        |\n",
    "   300000 |*+                                                                                                     +-|\n",
    "          |*                                                                                                        |\n",
    "          |*                                                                                                        |\n",
    "   200000 |*+                                                                                                     +-|\n",
    "          |**                                                                                                       |\n",
    "   100000 |-*                                                                                                     +-|\n",
    "          |  **                                                                                                     |\n",
    "          |    *************+                +                 +                 +                +                 |\n",
    "        0 +---------------------------------------------------------------------------------------------------------+\n",
    "          0                100              200               300               400              500               600\n",
    "                                                     Number of Individuals\n",
    "\n",
    "Please choose data cutoff.  Pick point right before the assymptote. A good starting cutoff might be 10% of the total number of individuals\n",
    "113\n",
    "At this point, all configuration information has been entered and dDocent may take several hours to run.\n",
    "It is recommended that you move this script to a background operation and disable terminal input and output.\n",
    "All data and logfiles will still be recorded.\n",
    "To do this:\n",
    "Press control and Z simultaneously\n",
    "Type 'bg' without the quotes and press enter\n",
    "Type 'disown -h' again without the quotes and press enter\n",
    "\n",
    "Now sit back, relax, and wait for your analysis to finish\n",
    "\n",
    "dDocent assembled 17509 sequences (after cutoffs) into 4012 contigs\n",
    "mawk: cannot open ./logfiles/VCFtools.log (No such file or directory)\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### move files to a folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "docentdir = '/home/lindb/eckertlab/BURT/fastqs_for_dDocent'\n",
    "# ls | grep -v \"GA\" | grep -v \"FL\" | grep -v \"AL\" | grep -v \"MS\" | grep -v \"TX\" | grep -v \"VA\" | grep -v \"LA\" | grep -v \"AR\" | grep -v \"NC\" | grep -v \"SC\" | grep -v \"PA\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
